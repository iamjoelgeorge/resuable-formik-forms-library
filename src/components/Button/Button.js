import React from 'react';

import styles from './Button.module.scss';

const Button = (props) => {
  const { label, type, onClick, ...rest } = props;

  return (
    <button className={styles.button} type={type} {...rest}>
      Submit
    </button>
  );
};

export default Button;
