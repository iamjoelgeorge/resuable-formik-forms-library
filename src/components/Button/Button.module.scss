@import '../../theme/variables.scss';

.button {
  border: none;
  padding: 10px;
  border-radius: 2px;
  background-color: $darkPurple;
  color: white;
  font-weight: bold;
  margin-top: 12px;

  &:disabled {
    background-color: rgb(89, 89, 89);
  }
}
