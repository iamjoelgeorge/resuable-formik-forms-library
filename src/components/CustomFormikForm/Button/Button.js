import React from 'react';

import PropTypes from 'prop-types';

import styles from './Button.module.scss';
import { joinClassNames } from '../../../utils/utils';
import {
  ExternalLinkIconWhite,
  ExternalLinkIconRed,
  ExternalLinkIconPurple,
} from '../../../constants/icons';

const Button = (props) => {
  const {
    variant,
    type,
    href,
    label,
    onClick,
    isDisabled,
    theme,
    showExternalLinkIcon,
    containerClass,
    formik,
    ...rest
  } = props;

  // const disableButton = type === 'submit' ? !(formik.dirty && formik.isValid) : isDisabled;
  const disableButton = type === 'submit' ? !formik.isValid : isDisabled;
  const goTo = !isDisabled ? href : '';

  const linkDisabledClass = isDisabled ? styles.linkDisabled : '';
  const linkAsButtonDisabledClass = isDisabled ? styles.linkButtonDisabled : '';

  const IconForLink = theme === 'purple' ? ExternalLinkIconPurple : ExternalLinkIconRed;

  const buttonClasses =
    theme === 'purple'
      ? joinClassNames([styles.button, styles.purpleTheme, containerClass])
      : joinClassNames([styles.button, styles.redTheme, containerClass]);

  const linkClasses =
    theme === 'purple'
      ? joinClassNames([styles.link, styles.purpleTheme, linkDisabledClass, containerClass])
      : joinClassNames([styles.link, styles.redTheme, linkDisabledClass, containerClass]);

  const linkAsButtonClasses = joinClassNames([
    buttonClasses,
    styles.linkAsButton,
    linkAsButtonDisabledClass,
  ]);

  const handleClickOnLink = (e) => {
    if (isDisabled) e.preventDefault();
  };

  const renderButton = () => {
    switch (variant) {
      case 'button':
        return (
          <button className={buttonClasses} type={type} disabled={disableButton} {...rest}>
            {label}
          </button>
        );

      case 'link':
        return (
          <a
            className={linkClasses}
            href={goTo}
            target='_blank'
            onClick={handleClickOnLink}
            rel='noreferrer'
          >
            <span>{label}</span>
            {showExternalLinkIcon && (
              <img className={styles.externalLinkIcon} src={IconForLink} alt='External Link' />
            )}
          </a>
        );

      case 'link_as_button':
        return (
          <a
            className={linkAsButtonClasses}
            href={goTo}
            target='_blank'
            onClick={handleClickOnLink}
            rel='noreferrer'
          >
            {label}
            {showExternalLinkIcon && (
              <img
                className={styles.externalLinkIcon}
                src={ExternalLinkIconWhite}
                alt='External Link'
              />
            )}
          </a>
        );

      default:
        return (
          <button
            className={buttonClasses}
            type={type}
            disabled={disableButton}
            onClick={onClick}
            {...rest}
          >
            {label}{' '}
            {showExternalLinkIcon && (
              <img
                className={styles.externalLinkIcon}
                src={ExternalLinkIconWhite}
                alt='External Link'
              />
            )}
          </button>
        );
    }
  };

  return <>{renderButton()}</>;
};

Button.propTypes = {
  variant: PropTypes.oneOf(['button', 'link', 'link_as_button']),
  type: PropTypes.oneOf(['submit', 'button']),
  href: PropTypes.string,
  label: PropTypes.string,
  onClick: PropTypes.func,
  isDisabled: PropTypes.bool,
  theme: PropTypes.oneOf(['red', 'purple']),
  showExternalLinkIcon: PropTypes.bool,
  containerClass: PropTypes.string,
};

Button.defaultProps = {
  variant: 'button',
  type: 'submit',
  href: '#',
  label: '',
  onClick: () => {},
  isDisabled: false,
  theme: 'purple',
  showExternalLinkIcon: false,
  containerClass: '',
};

export default Button;
