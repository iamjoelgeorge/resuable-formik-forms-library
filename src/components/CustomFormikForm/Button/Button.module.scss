@import '../../../theme/variables.scss';
@import '../../../theme/utils.scss';

* {
  box-sizing: border-box;
}

@mixin useButtonTheme($bgColor, $hoverBgColor, $focusBgColor) {
  background-color: $bgColor;

  &:hover:not(:disabled),
  &:active {
    cursor: pointer;
    background-color: $hoverBgColor;
  }

  &:focus {
    border-color: $focusBgColor;
  }
}

@mixin useLinkTheme(
  $defaultTextColor,
  $hoverColor,
  $borderBottomColorOnFocus,
  $borderBottomColorOnActive,
  $textColorOnActive
) {
  color: $defaultTextColor;

  span {
    border: 2px solid transparent;
    padding-bottom: 2px;
  }

  &:hover:not(&.linkDisabled) {
    color: $hoverColor;
  }

  &:focus {
    span {
      border-bottom-style: dashed;
      border-bottom-color: $borderBottomColorOnFocus;
    }
  }

  &:active {
    border-color: $borderBottomColorOnActive;
    color: $textColorOnActive;
  }
}

.button,
.link {
  border: 3px solid transparent;
  border-radius: 4px;
  font-size: em(16);
  font-weight: $fontWeightExtraBold;
  line-height: 19px;
  margin-top: 12px;
  padding: 0 em(15);
  width: 198px;
  height: 51px;
  text-align: center;
}

.button {
  color: $white;
  transition: 0.1s all ease-in-out;
  outline: none;

  &.purpleTheme {
    @include useButtonTheme($primary, $darkPurple, $brightPurple);
  }

  &.redTheme {
    @include useButtonTheme($tooltipRed, $buttonHover, $pinkOne);
  }

  &:disabled {
    background-color: $white;
    color: $tertiaryGrey;
    border-color: $dGrey;
  }
}

.linkAsButton {
  display: flex;
  justify-content: center;
  align-items: center;
  text-decoration: none;

  &.linkButtonDisabled {
    background-color: $white !important;
    cursor: not-allowed !important;
    color: $tertiaryGrey;
    border-color: $dGrey;
  }
}

.link {
  display: flex;
  justify-content: center;
  align-items: center;
  text-decoration: none;
  outline: none;

  &.purpleTheme {
    @include useLinkTheme($primary, $darkPurple, $darkPurple, $brightPurple, $darkPurple);
  }

  &.redTheme {
    @include useLinkTheme($tooltipRed, $buttonHover, $tooltipRed, $pinkOne, $buttonHover);
  }

  &.linkDisabled {
    color: $tertiaryGrey;
    pointer-events: none;

    .externalLinkIcon {
      -webkit-filter: invert(50%); /* safari 6.0 - 9.0 */
      filter: invert(50%);
    }
  }
}

.externalLinkIcon {
  margin-left: em(8);
}
