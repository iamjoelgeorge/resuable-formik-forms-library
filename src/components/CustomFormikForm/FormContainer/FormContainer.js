import React, { Children, cloneElement } from 'react';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';
import { validateInput } from '../../../utils/validateInput';

const FormContainer = (props) => {
  const { initialValues, children, validations, containerClass, onSubmit } = props;

  const validationObject = {};
  validations.forEach((item) => {
    validateInput(item, validationObject);
  });

  const validationSchema = Yup.object().shape(validationObject);

  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema}>
      {(formik) => {
        const childrenWithExtraProp = Children.map(children, (child) =>
          cloneElement(child, { formik }),
        );
        return <Form className={containerClass}>{childrenWithExtraProp}</Form>;
      }}
    </Formik>
  );
};

export default FormContainer;
