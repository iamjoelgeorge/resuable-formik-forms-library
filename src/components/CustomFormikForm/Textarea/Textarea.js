import React, { useState, useRef } from 'react';

import PropTypes from 'prop-types';
import { Field, useFormikContext } from 'formik';

import styles from './Textarea.module.scss';
import { commonProps, commonPropTypes } from '../../../constants/constants';
import { joinClassNames } from '../../../utils/utils';
import SlidingLabel from '../SlidingLabel/SlidingLabel';
import ErrorText from '../ErrorText/ErrorText';
import AdditionalInfo from '../AdditionalInfo/AdditionalInfo';

const Textarea = (props) => {
  const {
    name,
    label,
    placeholder,
    containerClass: customContainerClass,
    mainLabelTooltipBoxHeading,
    mainLabelTooltipBoxDescription,
    mainLabelTooltipBoxDescriptionElement,
    tooltipLink,
    tooltipLinkText,
    helpLinkText,
    helpLink,
    optionalText,
    isDisabled,
    isRequired,
    ...rest
  } = props;
  const [labelView, showLabelView] = useState(false);
  const [slideLabel, setSlideLabel] = useState(false);
  const textareaRef = useRef();

  const { values, errors, touched, handleBlur: formikHandleBlur } = useFormikContext();
  const inputValue = values[name];

  const userHasVisitedTheInputField = touched[name];
  const inputFieldHasErrors = errors[name];
  const addErrorClassesToLabelAndInput = !!userHasVisitedTheInputField && !!inputFieldHasErrors;

  const placeholderButtonClasses = joinClassNames([styles.textarea, styles.placeholderButton]);

  const handleBlurOnField = (e) => {
    formikHandleBlur(e);
    setSlideLabel(false);
    if (inputValue) showLabelView(true);
  };

  const handleClick = () => {
    showLabelView(false);

    setTimeout(() => {
      textareaRef?.current?.focus();
    }, 0);
  };

  const renderFieldView = () =>
    labelView && inputValue ? (
      <button
        type='button'
        data-testid={`label-view-${name}`}
        className={placeholderButtonClasses}
        tabIndex='0'
        onClick={handleClick}
      >
        {placeholder && !inputValue ? placeholder : inputValue}
      </button>
    ) : (
      <Field
        data-testid={`textarea-${name}`}
        innerRef={textareaRef}
        className={styles.textarea}
        as='textarea'
        name={name}
        placeholder={placeholder}
        disabled={isDisabled}
        onBlur={handleBlurOnField}
        {...rest}
      />
    );

  return (
    <div className={joinClassNames([styles.container, customContainerClass])}>
      <div className={styles.fieldContainer}>
        {renderFieldView()}
        {label && (
          <SlidingLabel
            label={label}
            htmlFor={name}
            inputEntered={!!inputValue || !!placeholder || slideLabel}
            inputIsDisabled={isDisabled}
            customClass={styles.label}
            showErrorStyle={addErrorClassesToLabelAndInput}
            tooltipBoxHeading={mainLabelTooltipBoxHeading}
            tooltipBoxDescription={mainLabelTooltipBoxDescription}
            tooltipBoxDescriptionElement={mainLabelTooltipBoxDescriptionElement}
            inputIsRequired={isRequired}
          />
        )}
      </div>

      <ErrorText containerClass={styles.errorContainer} fieldName={name} />

      <AdditionalInfo
        optionalText={optionalText}
        helpLinkText={helpLinkText}
        helpLink={helpLink}
        tooltipLinkText={tooltipLinkText}
        tooltipLink={tooltipLink}
      />
    </div>
  );
};

Textarea.propTypes = {
  ...commonPropTypes,
  label: PropTypes.string,
  placeholder: PropTypes.string,
};

Textarea.defaultProps = {
  label: PropTypes.string,
  placeholder: '',
  ...commonProps,
};

export default Textarea;
