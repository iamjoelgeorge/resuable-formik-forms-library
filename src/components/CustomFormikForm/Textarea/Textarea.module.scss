@import '../../../theme/variables.scss';
@import '../../../theme/utils.scss';

*,
*::before,
*::after {
  box-sizing: border-box;
}

.container {
  display: flex;
  justify-content: center;
  flex-direction: column;
  // padding: 18px 14px 0;
  margin-top: 20px;
  margin-bottom: 20px;
  outline: none;
  // border: 1px solid $lighterGrey;
  // border-radius: 4px;
  transition: 0.1s all ease-in-out;

  // &:focus {
  //   border-color: $brightPurple;
  //   border-style: dashed;
  // }

  // &:hover {
  //   border-color: $brightPurple;
  // }

  .fieldContainer {
    position: relative;
    height: 200px;
    cursor: pointer;
  }

  .slideLabel {
    top: em(25);
    font-size: 11px;
  }

  .textarea {
    resize: none;
    width: 100%;
    height: 100%;
    margin-top: 10px;
    padding-top: em(28);
    padding-left: em(14);
    border: 1px solid $lighterGrey;
    border-radius: 4px;
    outline: none;
    font-size: 13px;
    color: $primaryDark;
    cursor: pointer;
    overflow-y: scroll;

    &:focus {
      border-color: $darkPurple;

      & + label {
        @extend .slideLabel;
      }
    }

    &:hover {
      border-color: $brightPurple;
    }
  }

  .label {
    top: em(25) !important;
    left: em(14);
  }

  .placeholderButton {
    display: flex;
    justify-content: flex-start;
    align-items: flex-start;
    text-align: left;
    background-color: $white;
    position: relative;
    padding-bottom: em(20);
    line-height: 18px;

    &:focus {
      border-color: $brightPurple;
      border-style: dashed;
    }
  }
}
