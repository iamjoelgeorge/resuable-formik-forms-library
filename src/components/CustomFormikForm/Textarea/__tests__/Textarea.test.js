import React from 'react';

import { render, cleanup, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import Button from '../../Button/Button';
import FormContainer from '../../FormContainer/FormContainer';
import Textarea from '../Textarea';

const defaultProps = {
  name: 'description',
  label: 'description',
  placeholder: '',
  containerClass: '',
  labelTooltipBoxHeading: '',
  labelTooltipBoxDescription: '',
  labelTooltipBoxDescriptionElement: null,
  tooltipLink: '',
  tooltipLinkText: '',
  helpLinkText: '',
  helpLink: '',
  optionalText: '',
  isDisabled: false,
  isRequired: false,
  formik: {
    errors: {
      description: '',
    },
    touched: { description: false },
    values: { description: '' },
  },
};

const initialValues = {
  description: '',
};

const defaultValidations = [];
const handleSubmit = jest.fn();

afterEach(cleanup);

describe('[Component]: Texxtarea', () => {
  it('should match the Snapshot', () => {
    const tree = renderer
      .create(
        <FormContainer
          initialValues={initialValues}
          validations={defaultValidations}
          onSubmit={handleSubmit}
        >
          <Textarea {...defaultProps} />
        </FormContainer>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the textarea correctly', () => {
    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <Textarea {...defaultProps} />
      </FormContainer>,
    );

    expect(getByTestId('textarea-description')).toBeTruthy();
  });

  it('should render a disabled input', () => {
    const props = {
      ...defaultProps,
      isDisabled: true,
    };

    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
      </FormContainer>,
    );

    expect(getByTestId('textarea-description')).toHaveAttribute('disabled');
  });

  it('should render the placeholder button', () => {
    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <Textarea {...defaultProps} />
      </FormContainer>,
    );

    userEvent.type(getByTestId('textarea-description'), 'John');
    getByTestId('textarea-description').blur();
    expect(getByTestId('label-view-description')).toBeTruthy();
  });

  it('should render the optional text', () => {
    const props = {
      ...defaultProps,
      optionalText: 'Test Optional text',
    };

    const { getByText } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
      </FormContainer>,
    );

    expect(getByText('Test Optional text')).toBeTruthy();
  });

  it('should render the help link', () => {
    const props = {
      ...defaultProps,
      helpLinkText: 'Test help link',
      helpLink: 'https://virginaustralia.com/',
    };

    const { getByText } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
      </FormContainer>,
    );

    expect(getByText('Test help link')).toBeTruthy();
  });

  it('should render the tooltip link', () => {
    const props = {
      ...defaultProps,
      tooltipLinkText: 'Test tooltip link',
      tooltipLink: 'https://virginaustralia.com/',
    };

    const { getByText } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
      </FormContainer>,
    );

    expect(getByText('Test tooltip link')).toBeTruthy();
  });

  it('should show the error component for a required input', async () => {
    const props = {
      ...defaultProps,
      isRequired: true,
    };

    const validations = [
      {
        name: 'description',
        type: 'string',
        isRequired: true,
        message: 'Please add a comment.',
      },
    ];

    const { getByText, getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={validations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
        <Button label='Submit' />
      </FormContainer>,
    );

    const submitButton = getByText('Submit');
    userEvent.click(submitButton);

    await waitFor(() => {
      expect(getByTestId('error-description')).toHaveTextContent('Please add a comment');
    });
  });

  it('should show the error component for maximum characters allowed', async () => {
    const props = {
      ...defaultProps,
      isRequired: true,
    };

    const validations = [
      {
        name: 'description',
        type: 'string',
        isRequired: false,
        maxChars: { num: 5, message: 'You can enter only upto 5 characters' },
      },
    ];

    const { getByText, getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={validations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
        <Button label='Submit' />
      </FormContainer>,
    );

    const textarea = getByTestId('textarea-description');
    const submitButton = getByText('Submit');

    userEvent.type(textarea, 'Less');
    userEvent.click(submitButton);

    await waitFor(() => {
      const labelViewButton = getByTestId('label-view-description');
      userEvent.click(labelViewButton);
    });

    userEvent.clear(textarea);
    userEvent.type(textarea, 'valid input');

    await waitFor(() => {
      expect(getByTestId('error-description')).toHaveTextContent(
        'You can enter only upto 5 characters',
      );
    });
  });

  it('should show the error component for minimum characters required', async () => {
    const props = {
      ...defaultProps,
      isRequired: true,
    };

    const validations = [
      {
        name: 'description',
        type: 'string',
        isRequired: false,
        minChars: { num: 5, message: 'Please enter at least 5 characters' },
      },
    ];

    const { getByText, getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={validations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
        <Button label='Submit' />
      </FormContainer>,
    );

    const textarea = getByTestId('textarea-description');
    const submitButton = getByText('Submit');

    userEvent.type(textarea, 'valid input');
    userEvent.click(submitButton);

    await waitFor(() => {
      const labelViewButton = getByTestId('label-view-description');
      userEvent.click(labelViewButton);
    });

    userEvent.clear(textarea);
    userEvent.type(textarea, 'jn');

    await waitFor(() => {
      expect(getByTestId('error-description')).toHaveTextContent(
        'Please enter at least 5 characters',
      );
    });
  });

  it('should show the error component for minimum and maximum characters', async () => {
    const props = {
      ...defaultProps,
      isRequired: true,
    };

    const validations = [
      {
        name: 'description',
        type: 'string',
        isRequired: false,
        minChars: { num: 5, message: 'Please enter at least 5 characters' },
        maxChars: { num: 10, message: 'You can enter only upto 10 characters' },
      },
    ];

    const { getByText, getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={validations}
        onSubmit={handleSubmit}
      >
        <Textarea {...props} />
        <Button label='Submit' />
      </FormContainer>,
    );

    const textarea = getByTestId('textarea-description');
    const submitButton = getByText('Submit');

    userEvent.type(textarea, 'valid input');
    userEvent.click(submitButton);

    await waitFor(() => {
      const labelViewButton = getByTestId('label-view-description');
      userEvent.click(labelViewButton);
    });

    userEvent.clear(textarea);
    userEvent.type(textarea, 'jn');

    await waitFor(() => {
      expect(getByTestId('error-description')).toHaveTextContent(
        'Please enter at least 5 characters',
      );
    });

    userEvent.clear(textarea);
    userEvent.type(textarea, 'Entering more than 10 characters');

    await waitFor(() => {
      expect(getByTestId('error-description')).toHaveTextContent(
        'You can enter only upto 10 characters',
      );
    });
  });
});
