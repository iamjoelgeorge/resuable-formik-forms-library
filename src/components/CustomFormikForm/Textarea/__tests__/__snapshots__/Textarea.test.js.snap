// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component]: Texxtarea should match the Snapshot 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="container "
  >
    <div
      className="fieldContainer"
    >
      <textarea
        className="textarea"
        data-testid="textarea-description"
        disabled={false}
        name="description"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder=""
        value=""
      />
      <div
        className="container  label"
      >
        <label
          className="label"
          htmlFor="description"
        >
          description
        </label>
      </div>
    </div>
    
    
    
  </div>
</form>
`;
